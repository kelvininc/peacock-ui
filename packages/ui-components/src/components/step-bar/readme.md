# _<kv-step-bar>_



<!-- Auto Generated Below -->


## Usage

### Angular

```html
<!-- Default -->
<kv-step-bar steps={STEPS} currentStep={1} progressPercentage={50}></kv-step-bar>
```


### React

```tsx
import React from 'react';

import { KvStepBar } from '@kelvininc/react-ui-components';

export const StepBarExample: React.FC = () => {
	const steps = [
		{
			enabled: true,
			active: true
		},
		{
			enabled: true,
			active: true
		},
		{
			enabled: false
		}
	];

	return (
		<>
			{/*-- Default --*/}
			<KvStepBar steps={steps} currentStep={1} progressPercentage={50}></KvStepBar>
		</>
	);
};
```



## Properties

| Property                          | Attribute             | Description                                                                                  | Type               | Default       |
| --------------------------------- | --------------------- | -------------------------------------------------------------------------------------------- | ------------------ | ------------- |
| `currentStep` _(required)_        | `current-step`        | (required) Defines the current step index                                                    | `number`           | `undefined`   |
| `hasError`                        | `has-error`           | (optional) Defines if the progress bar should be in an error state                           | `boolean`          | `undefined`   |
| `label`                           | `label`               | (optional) Defines the label to display next to the step counter (defaults to: "Progress: ") | `string`           | `'Progress:'` |
| `progressPercentage` _(required)_ | `progress-percentage` | (required) Defines the percentage of steps completed                                         | `number`           | `undefined`   |
| `steps` _(required)_              | --                    | (required) Defines the steps array to render                                                 | `IStepIndicator[]` | `undefined`   |


## Events

| Event         | Description                                      | Type                  |
| ------------- | ------------------------------------------------ | --------------------- |
| `stepClicked` | Fires when a step is clicked and emits the index | `CustomEvent<number>` |


## Dependencies

### Depends on

- [kv-step-progress-bar](../step-progress-bar)
- [kv-step-indicator](../step-indicator)

### Graph
```mermaid
graph TD;
  kv-step-bar --> kv-step-progress-bar
  kv-step-bar --> kv-step-indicator
  style kv-step-bar fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------


