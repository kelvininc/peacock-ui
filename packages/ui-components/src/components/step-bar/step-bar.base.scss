@import '../../assets/styles/globals';

/**
* @prop --step-label-color: The color of the label.
* @prop --step-counter-color: The color of the step counter.
*/

.label-container {
	.label {
		@include kv-font-label-small-semibold;
		color: var(--step-label-color);
	}

	.step-counter {
		@include kv-font-label-small-regular;
		color: var(--step-counter-color);

		margin-left: $spacing;
	}
}

.progress-bar-container {
	margin-top: $spacing;
}
