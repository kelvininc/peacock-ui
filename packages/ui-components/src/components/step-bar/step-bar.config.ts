export const DEFAULT_LABEL = 'Progress:';
