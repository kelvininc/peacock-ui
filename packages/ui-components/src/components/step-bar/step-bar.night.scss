@import './step-bar.base';

:host {
	--step-label-color: #{kv-color('neutral-4')};
	--step-counter-color: #{kv-color('neutral-1')};
}
