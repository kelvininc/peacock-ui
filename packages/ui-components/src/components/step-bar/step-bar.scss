@import '../../assets/styles/globals';

/**
* @prop --step-label-color: The color of the label.
* @prop --step-counter-color: The color of the step counter.
*/
$progress-bar-height: 2px;

:host {
	--step-label-color: #{kv-color('neutral-0')};
	--step-counter-color: #{kv-color('neutral-1')};
}

.label-container {
	.label {
		@include kv-font-label-small-semibold;
		color: var(--step-label-color);
	}

	.step-counter {
		@include kv-font-label-small-regular;

		color: var(--step-counter-color);
		margin-left: $spacing;
	}
}

.progress-bar-container {
	margin-top: $spacing;

	kv-step-progress-bar {
		--progress-bar-height: #{$progress-bar-height};
		--progress-height: 14px;
	}

	kv-step-indicator {
		--indicator-radius: 14px;
	}

	.step-indicator-container {
		display: flex;
		align-items: center;

		.error-bar-filler {
			position: absolute;
			height: $progress-bar-height;
			background-color: kv-color('error');
			z-index: 1;
		}

		.step-indicator {
			z-index: 2;
		}
	}
}
