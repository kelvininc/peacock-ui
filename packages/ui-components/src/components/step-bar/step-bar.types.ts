import { EventEmitter } from '@stencil/core';
import { IStepIndicator } from '../step-indicator/step-indicator.types';

export interface IStepBar {
	/** (optional) Defines the label to display next to the step counter (defaults to: "Progress: ") */
	label?: string;
	/** (required) Defines the steps array to render */
	steps: IStepIndicator[];
	/** (required) Defines the current step index */
	currentStep: number;
	/** (required) Defines the percentage of steps completed */
	progressPercentage: number;
	/** (optional) Defines if the progress bar should be in an error state */
	hasError: boolean;
}

export interface IStepBarEvents {
	/** Fires when a step is clicked and emits the index */
	stepClicked: EventEmitter<number>;
}
