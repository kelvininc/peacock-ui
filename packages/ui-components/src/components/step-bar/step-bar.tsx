import { Component, Event, EventEmitter, Host, Prop, h } from '@stencil/core';
import { IStepIndicator } from '../step-indicator/step-indicator.types';
import { IStepBar, IStepBarEvents } from './step-bar.types';

@Component({
	tag: 'kv-step-bar',
	styleUrls: {
		night: 'step-bar.night.scss'
	},
	shadow: true
})
export class KvStepBar implements IStepBar, IStepBarEvents {
	/** @inheritdoc */
	@Prop({ reflect: true }) label: string = 'Progress:';
	/** @inheritdoc */
	@Prop({ reflect: true }) steps!: IStepIndicator[];
	/** @inheritdoc */
	@Prop({ reflect: true }) currentStep!: number;
	/** @inheritdoc */
	@Prop({ reflect: true }) progressPercentage!: number;
	/** @inheritdoc */
	@Prop({ reflect: true }) hasError: boolean;

	/** @inheritdoc */
	@Event() stepClicked: EventEmitter<number>;

	private onStepClick = ({ detail }: CustomEvent<number>) => {
		this.stepClicked.emit(detail);
	};

	render() {
		return (
			<Host>
				<div class="label-container">
					<span class="label">{this.label}</span>
					<span class="step-counter">
						{this.currentStep + 1}/{this.steps.length} Steps
					</span>
				</div>
				<div class="progress-bar-container">
					<kv-step-progress-bar progressPercentage={this.progressPercentage} hasError={this.hasError}>
						{this.steps.map(({ active, enabled, hasError }, idx) => (
							<kv-step-indicator stepKey={idx} enabled={enabled} active={active} hasError={hasError} onIndicatorClicked={this.onStepClick} />
						))}
					</kv-step-progress-bar>
				</div>
			</Host>
		);
	}
}
