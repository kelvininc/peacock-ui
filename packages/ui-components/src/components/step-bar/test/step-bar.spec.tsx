import { newSpecPage, SpecPage } from '@stencil/core/testing';
import { h } from '@stencil/core';
import { STEPS_ERROR_MOCK, STEPS_SUCCESS_MOCK } from './step-bar.mock';
import { KvStepBar } from '../step-bar';

describe('KvStepBar (unit tests)', () => {
	let page: SpecPage;
	let component: KvStepBar;

	describe('when rendering with required props', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvStepBar],
				template: () => <kv-step-bar steps={STEPS_SUCCESS_MOCK} currentStep={1} progressPercentage={50}></kv-step-bar>
			});
			component = page.rootInstance;
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});

		it('should set the correct steps array', () => {
			expect(component.steps).toEqual([
				{
					enabled: true,
					active: true
				},
				{
					enabled: true,
					active: true
				},
				{
					enabled: false
				}
			]);
		});
	});

	describe('when rendering with required props and `hasError` true', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvStepBar],
				template: () => <kv-step-bar steps={STEPS_ERROR_MOCK} currentStep={1} progressPercentage={50}></kv-step-bar>
			});
			component = page.rootInstance;
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});

		it('should set the correct steps array', () => {
			expect(component.steps).toEqual([
				{
					enabled: true,
					active: true,
					hasError: true
				},
				{
					enabled: true,
					active: true,
					hasError: true
				},
				{
					enabled: false
				}
			]);
		});
	});
});
