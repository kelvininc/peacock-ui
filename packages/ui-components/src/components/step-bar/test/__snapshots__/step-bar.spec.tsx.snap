// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KvStepBar (unit tests) when rendering with required props and \`hasError\` true should match the snapshot 1`] = `
<kv-step-bar current-step="1" label="Progress:" progress-percentage="50">
  <mock:shadow-root>
    <div class="label-container">
      <span class="label">
        Progress:
      </span>
      <span class="step-counter">
        2/3 Steps
      </span>
    </div>
    <div class="progress-bar-container">
      <kv-step-progress-bar progresspercentage="50">
        <kv-step-indicator active="" enabled="" haserror="" stepkey="0"></kv-step-indicator>
        <kv-step-indicator active="" enabled="" haserror="" stepkey="1"></kv-step-indicator>
        <kv-step-indicator stepkey="2"></kv-step-indicator>
      </kv-step-progress-bar>
    </div>
  </mock:shadow-root>
</kv-step-bar>
`;

exports[`KvStepBar (unit tests) when rendering with required props should match the snapshot 1`] = `
<kv-step-bar current-step="1" label="Progress:" progress-percentage="50">
  <mock:shadow-root>
    <div class="label-container">
      <span class="label">
        Progress:
      </span>
      <span class="step-counter">
        2/3 Steps
      </span>
    </div>
    <div class="progress-bar-container">
      <kv-step-progress-bar progresspercentage="50">
        <kv-step-indicator active="" enabled="" stepkey="0"></kv-step-indicator>
        <kv-step-indicator active="" enabled="" stepkey="1"></kv-step-indicator>
        <kv-step-indicator stepkey="2"></kv-step-indicator>
      </kv-step-progress-bar>
    </div>
  </mock:shadow-root>
</kv-step-bar>
`;
